src/Workspaces/Core/Portable/Utilities/SpellChecker.cs

```diff
@@ -97,68 +96,85 @@ public CacheResult(string candidate, bool areSimilar, double similarityWeight)
+        private bool _substringsAreSimilar;
+
+        private static readonly object s_poolGate = new object();
+        private static readonly Stack<WordSimilarityChecker> s_pool = new Stack<WordSimilarityChecker>();
+
+        public static WordSimilarityChecker Allocate(string text, bool substringsAreSimilar)
+        {
+            WordSimilarityChecker checker;
+            lock (s_poolGate)
+            {
+                checker = s_pool.Count > 0
+                    ? s_pool.Pop()
+                    : new WordSimilarityChecker();
+            }
+
+            checker.Initialize(text, substringsAreSimilar);
+            return checker;
+        }
+
+        private WordSimilarityChecker()
+        {
+        }
 
-        public WordSimilarityChecker(string text, bool substringsAreSimilar)
+        private void Initialize(string text, bool substringsAreSimilar)
         {
             _source = text ?? throw new ArgumentNullException(nameof(text));
             _threshold = GetThreshold(_source);
             _editDistance = new EditDistance(text);
             _substringsAreSimilar = substringsAreSimilar;
         }
 
-        public void Dispose()
+        public void Free()
         {
             _editDistance?.Dispose();
+            _source = null;
             _editDistance = null;
+            _lastAreSimilarResult = default(CacheResult);
+            s_pool.Push(this);
         }
 
         public static bool AreSimilar(string originalText, string candidateText)
-        {
-            return AreSimilar(originalText, candidateText, substringsAreSimilar: false);
-        }
+            => AreSimilar(originalText, candidateText, substringsAreSimilar: false);
 
         public static bool AreSimilar(string originalText, string candidateText, bool substringsAreSimilar)
-        {
-            return AreSimilar(originalText, candidateText, substringsAreSimilar, out var unused);
-        }
+            => AreSimilar(originalText, candidateText, substringsAreSimilar, out var unused);
 
         public static bool AreSimilar(string originalText, string candidateText, out double similarityWeight)
         {
             return AreSimilar(
-                originalText, candidateText, 
+                originalText, candidateText,
                 substringsAreSimilar: false, similarityWeight: out similarityWeight);
         }
-        
+
         /// <summary>
         /// Returns true if 'originalText' and 'candidateText' are likely a misspelling of each other.
         /// Returns false otherwise.  If it is a likely misspelling a similarityWeight is provided
         /// to help rank the match.  Lower costs mean it was a better match.
         /// </summary>
         public static bool AreSimilar(string originalText, string candidateText, bool substringsAreSimilar, out double similarityWeight)
         {
-            using (var checker = new WordSimilarityChecker(originalText, substringsAreSimilar))
```

Comment:
See below... i struggled with that idea, but in the end went with teh compiler pattern which i think makes it more clean that the object *is actually being returned*.